### PR TITLE
[ASDisplayNode] Fix infinite layout loop

### PR DESCRIPTION
**Pull request by nguyenhuy**
The problem was introduced by #428. It will occur if a node does either of the followings:
1. Keeps using a stale pending layout over a calculated layout
2. Doesn't update the next layout's version after calling `_setNeedsLayoutFromAbove`.

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/09a91e3435a065ab8dacb50c8a8bafe7/raw/dfb68e245b4730ac99fd7662a04806f1a25a5fb2/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by nguyenhuy** [comment]
@Adlai-Holler Yes, it indeed sounds like a cleaner approach. However, there are a couple of problems with it:
- Calling `-setNeedsLayout` on the layer is only true for the node that originates the call. Since `_setNeedsLayoutFromAbove` is recursive, all other higher up nodes need to invalidate their layouts, so `[self setNeedsLayout]` is needed. This can be solved by passing the original node as a parameter and compare against `self`.
- Another bigger problems is that, some classes override `-setNeedsLayout` to add custom logics. Metadata node for example needs to reset its animation flags as soon as the current layout is invalid. Not calling `[self setNeedsLayout]` in this case is an API breaking change.

**Comment by appleguy** [comment]
Should we revert the layoutVersion change for now? It's definitely necessary (currently) to call setNeedsLayout while ascending the tree, so we should be careful to align any changes with the current expectations of the code.

File: Source/ASDisplayNode+Layout.mm [hunk 1]
```diff
@@ -352,8 +352,10 @@ - (void)_locked_measureNodeWithBoundsIfNecessary:(CGRect)bounds
     ASLayout *layout = [self calculateLayoutThatFits:constrainedSize
                                     restrictedToSize:self.style.size
                                 relativeToParentSize:boundsSizeForLayout];
-    
     nextLayout = std::make_shared<ASDisplayNodeLayout>(layout, constrainedSize, boundsSizeForLayout, version);
+    // Now that the constrained size of pending layout might have been reused, the layout is useless
+    // Release it and any orphaned subnodes it retains
+    _pendingDisplayNodeLayout = nullptr;
```
Comment (appleguy) [inline Source/ASDisplayNode+Layout.mm]: Does this effectively replace the line I commented on here? https://github.com/TextureGroup/Texture/pull/428/files#r126296449

If so, we should probably replace the line at that same location. I think it should not have been removed originally.

Comment (nguyenhuy) [inline Source/ASDisplayNode+Layout.mm]: Yes, more or less. The main difference is that here only stale or inapplicable pending layouts are cleared, so it can happen that at a given time, both calculated and pending layouts point to a valid layout (and thus can be interchangeable). With the other original line, all pending layouts are cleared, including valid/applied ones.

Comment (Adlai-Holler) [inline Source/ASDisplayNode+Layout.mm]: Why is this line necessary @nguyenhuy ?

Comment (nguyenhuy) [inline Source/ASDisplayNode+Layout.mm]: Either of this line or the new layout version check [above](https://github.com/TextureGroup/Texture/pull/455/files#diff-71ea742a29543cb7e08c3dffe2a2508eR308) fixes part of the problem. The reason I included it is that at this point, the pending layout can be stale ([here](https://github.com/TextureGroup/Texture/pull/455/files#diff-71ea742a29543cb7e08c3dffe2a2508eR338)) and it's never going to be applied. And yet, it still holds onto orphaned subnodes that were either never inserted or removed by newer layouts (check out `-[ASLayout retainSublayoutLayoutElements]`).

Comment (nguyenhuy) [inline Source/ASDisplayNode+Layout.mm]: It is true that in case the pending layout can't be applied because of a bounds mismatch, we can keep it around. I'm fine with adding another flag [here](https://github.com/TextureGroup/Texture/pull/455/files#diff-71ea742a29543cb7e08c3dffe2a2508eR340) to handle that edge.

However, as said above, in case `_pendingDisplayNodeLayout->version < _layoutVersion`, it's quite important to get rid of it.

Comment (Adlai-Holler) [inline Source/ASDisplayNode+Layout.mm]: Sorry I must be missing something big. Why is it important to get rid of it? What's the benefit?

Comment (Adlai-Holler) [inline Source/ASDisplayNode+Layout.mm]: Oh there's a comment between those two that I missed. OK that makes sense, to release those nodes. We should have a common pattern that in one step checks if a given layout is valid, and discards it if not, and use that everywhere, but yes I'm on the same page about this line now.

Comment (nguyenhuy) [inline Source/ASDisplayNode+Layout.mm]: 👍

**Comment by nguyenhuy** [comment]
@Adlai-Holler Agreed!

I reckon the long-term solution would be to always layout from root.

**Comment by nguyenhuy** [comment]
I'm gonna merge this PR and run an internal test. If we encounter another problem, I think we should really consider to revert the original PR as @appleguy suggested. Thanks everyone!
